File: java-advent-of-code/src/main/java/com/mycompany/advent_of_code_java/Advent_of_code_2025_004_002.java
package com.mycompany.advent_of_code_java;

import module java.base;

public class Advent_of_code_2025_004_002 {

    public static void main(String[] args) {
        String Input_path = (args.length >= 1)
                ? args[0]
                : "E:\\documents-container\\advent-of-code\\java-advent-of-code\\inputs\\2025-004\\input.txt";

        try {
            long Total_removed = Solve(Input_path);
            System.out.println(Total_removed);
        } catch (IOException e) {
            System.err.println("Failed to read input: " + Input_path);
            e.printStackTrace(System.err);
            System.exit(1);
        } catch (IllegalArgumentException e) {
            System.err.println("Invalid input: " + e.getMessage());
            System.exit(2);
        }
    }

    public static long Solve(String Input_path) throws IOException {
        List<String> Lines = Files.readAllLines(Path.of(Input_path));
        if (Lines.isEmpty()) return 0;

        int Rows = Lines.size();
        int Cols = Lines.get(0).length();

        for (int r = 0; r < Rows; r++) {
            if (Lines.get(r).length() != Cols) {
                throw new IllegalArgumentException("Grid is not rectangular (line " + (r + 1) + ").");
            }
        }

        char[][] Grid = new char[Rows][Cols];
        int[][] Adjacent_rolls = new int[Rows][Cols];

        long Total_rolls = 0;

        for (int r = 0; r < Rows; r++) {
            String Line = Lines.get(r);
            for (int c = 0; c < Cols; c++) {
                char ch = Line.charAt(c);
                if (ch != '@' && ch != '.') {
                    throw new IllegalArgumentException(
                            "Unexpected character '" + ch + "' at row " + (r + 1) + ", col " + (c + 1) + "."
                    );
                }
                Grid[r][c] = ch;
                if (ch == '@') Total_rolls++;
            }
        }

        int[] Dr = {-1,-1,-1, 0, 0, 1, 1, 1};
        int[] Dc = {-1, 0, 1,-1, 1,-1, 0, 1};

        // Compute initial adjacent-roll counts.
        for (int r = 0; r < Rows; r++) {
            for (int c = 0; c < Cols; c++) {
                if (Grid[r][c] != '@') continue;

                int Count = 0;
                for (int k = 0; k < 8; k++) {
                    int rr = r + Dr[k];
                    int cc = c + Dc[k];
                    if (rr < 0 || rr >= Rows || cc < 0 || cc >= Cols) continue;
                    if (Grid[rr][cc] == '@') Count++;
                }
                Adjacent_rolls[r][c] = Count;
            }
        }

        ArrayDeque<Integer> Queue = new ArrayDeque<>();
        boolean[][] Enqueued = new boolean[Rows][Cols];

        // Seed queue with initially removable rolls.
        for (int r = 0; r < Rows; r++) {
            for (int c = 0; c < Cols; c++) {
                if (Grid[r][c] == '@' && Adjacent_rolls[r][c] < 4) {
                    Queue.addLast(Encode(r, c, Cols));
                    Enqueued[r][c] = true;
                }
            }
        }

        long Removed = 0;

        while (!Queue.isEmpty()) {
            int Code = Queue.removeFirst();
            int r = Code / Cols;
            int c = Code % Cols;

            // It might have been removed already or become non-roll; skip if so.
            if (Grid[r][c] != '@') continue;

            // Remove it.
            Grid[r][c] = '.';
            Removed++;

            // Update neighbors: each neighboring roll loses one adjacent roll.
            for (int k = 0; k < 8; k++) {
                int rr = r + Dr[k];
                int cc = c + Dc[k];
                if (rr < 0 || rr >= Rows || cc < 0 || cc >= Cols) continue;

                if (Grid[rr][cc] == '@') {
                    Adjacent_rolls[rr][cc]--;

                    // If it just became removable, enqueue it.
                    if (Adjacent_rolls[rr][cc] < 4 && !Enqueued[rr][cc]) {
                        Queue.addLast(Encode(rr, cc, Cols));
                        Enqueued[rr][cc] = true;
                    }
                }
            }
        }

        // Removed is the maximum removable by repeatedly removing accessible rolls.
        // (Order does not reduce removability here; any eligible roll can be removed.)
        return Removed;
    }

    private static int Encode(int r, int c, int Cols) {
        return r * Cols + c;
    }
}

File: python/Advent_of_code_2025_004_002.py
import sys
from pathlib import Path
from collections import deque

def solve(input_path_str: str) -> int:
    input_path = Path(input_path_str)
    
    if not input_path.exists():
        raise FileNotFoundError(f"File not found: {input_path}")

    # Read lines and initialize grid
    lines = [line.strip() for line in input_path.read_text(encoding='utf-8').splitlines() if line.strip()]
    if not lines:
        return 0

    rows = len(lines)
    cols = len(lines[0])
    
    grid = []
    adjacent_counts = [[0] * cols for _ in range(rows)]
    
    # 1. Build grid and validate
    for r, line in enumerate(lines):
        if len(line) != cols:
            raise ValueError(f"Grid is not rectangular (line {r + 1}).")
        grid.append(list(line))

    # 8-direction offsets
    dr = [-1, -1, -1,  0, 0,  1, 1, 1]
    dc = [-1,  0,  1, -1, 1, -1, 0, 1]

    # 2. Precompute initial adjacency counts
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] != '@':
                continue
            
            count = 0
            for k in range(8):
                rr, cc = r + dr[k], c + dc[k]
                if 0 <= rr < rows and 0 <= cc < cols:
                    if grid[rr][cc] == '@':
                        count += 1
            adjacent_counts[r][c] = count

    # 3. Seed the queue with rolls that are initially removable
    queue = deque()
    enqueued = [[False] * cols for _ in range(rows)]

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '@' and adjacent_counts[r][c] < 4:
                queue.append((r, c))
                enqueued[r][c] = True

    removed_count = 0

    # 4. Simulation loop (BFS)
    while queue:
        r, c = queue.popleft()

        # Skip if already removed (though 'enqueued' guard usually prevents this)
        if grid[r][c] != '@':
            continue

        # Remove the roll
        grid[r][c] = '.'
        removed_count += 1

        # Notify neighbors
        for k in range(8):
            rr, cc = r + dr[k], c + dc[k]
            if 0 <= rr < rows and 0 <= cc < cols:
                if grid[rr][cc] == '@':
                    adjacent_counts[rr][cc] -= 1

                    # If neighbor becomes removable, add to queue
                    if adjacent_counts[rr][cc] < 4 and not enqueued[rr][cc]:
                        queue.append((rr, cc))
                        enqueued[rr][cc] = True

    return removed_count

def main():
    default_path = r"E:\documents-container\advent-of-code\java-advent-of-code\inputs\2025-004\input.txt"
    input_path_str = sys.argv[1] if len(sys.argv) >= 2 else default_path
    
    try:
        result = solve(input_path_str)
        print(result)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()